File: 2023/go/day02a/day02a.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	TOTAL_RED   = 12
	TOTAL_GREEN = 13
	TOTAL_BLUE  = 14
)

func usage(progname string) {
	log.Fatalf("usage: %s <file>\n", progname)
	os.Exit(1)
}

func process(contents string) uint32 {
	result := uint32(0)
	lines := strings.Split(contents, "\n")
	for _, line := range lines {
		if len(line) == 0 {
			break
		}
		parts := strings.Split(line, ": ")
		gamePart, drawsPart := parts[0], parts[1]
		parts = strings.Split(gamePart, " ")
		gameNumStr := parts[1]
		gameNum, err := strconv.ParseUint(gameNumStr, 10, 32)
		if err != nil {
			panic(err)
		}
		valid := true
		for _, drawPart := range strings.Split(drawsPart, "; ") {
			for _, colorAmount := range strings.Split(drawPart, ", ") {
				parts = strings.Split(colorAmount, " ")
				amountStr, color := parts[0], parts[1]
				amount, err := strconv.ParseUint(amountStr, 10, 32)
				if err != nil {
					panic(err)
				}
				switch color {
				case "red":
					if amount > TOTAL_RED {
						valid = false
					}
				case "green":
					if amount > TOTAL_GREEN {
						valid = false
					}
				case "blue":
					if amount > TOTAL_BLUE {
						valid = false
					}
				default:
					panic("unknown color")
				}
			}
		}
		if valid {
			result += uint32(gameNum)
		}
	}
	return result
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}
	filename := os.Args[1]
	buf, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}
	contents := string(buf[:])
	result := process(contents)
	fmt.Printf("result = %d\n", result)
}

File: 2023/r/day02a/day02a.r
#!/usr/bin/env Rscript

PROG_NAME <- "day02a.r"

TOTAL_RED <- 12
TOTAL_GREEN <- 13
TOTAL_BLUE <- 14

usage <- function() {
    cat("usage:", PROG_NAME, "<file>\n")
    quit(status=1)
}

process <- function(filename) {
    result <- 0
    infile <- file(filename, "r")
    while ( length(line <- readLines(infile, 1)) > 0 ) {
        if ( !is.null(strsplit(line, ": ")[[1]]) ) {
            line_parts <- strsplit(line, ": ")[[1]]
            game_part <- line_parts[[1]]
            draws_part <- line_parts[[2]]
            if ( !is.null(strsplit(game_part, ' ')[[1]]) ) {
                game_parts <- strsplit(game_part, ' ')[[1]]
                game_num_str <- game_parts[[2]]
                game_num <- as.numeric(game_num_str)
                valid <- TRUE
                for ( draw_str in strsplit(draws_part, "; ")[[1]] ) {
                    for ( color_amount in strsplit(draw_str, ", ")[[1]] ) {
                        color_amount_parts <- strsplit(color_amount, ' ')[[1]]
                        amount_str <- color_amount_parts[[1]]
                        color <- color_amount_parts[[2]]
                        amount <- as.numeric(amount_str)
                        switch (color,
                            "red" = {
                                if ( amount > TOTAL_RED ) {
                                    valid <- FALSE
                                }
                            },
                            "green" = {
                                if ( amount > TOTAL_GREEN ) {
                                    valid <- FALSE
                                }
                            },
                            "blue" = {
                                if ( amount > TOTAL_BLUE ) {
                                    valid <- FALSE
                                }
                            },
                            stop("unknown color")
                        )
                    }
                }
                if ( valid ) {
                    result <- result + game_num
                }
            }
        }
    }
    result
}

main <- function() {
    args <- commandArgs(trailingOnly=TRUE)
    if ( length(args) < 1 ) {
        usage()
    }
    filename <- args[1]
    result <- process(filename)
    cat("result =", result, "\n")
}

if (!interactive()) {
    main()
}
